### PR TITLE
DH-1398 Enable SSO on circle

### PR DESCRIPTION
**Pull request by feedmypixel**
**TL;DR** This work enables `SSO` on CircleCi. 

This is part 1 of an amount of [Yak shaving](https://en.wiktionary.org/wiki/yak_shaving) required to test different user permissions in Acceptance tests.

I have created [uktrade/mock-sso](https://github.com/uktrade/mock-sso) nodeJs app. This app is built into a docker container via its [automated build](https://hub.docker.com/r/ukti/mock-sso/). This docker is then used in our [Acceptance tests](https://circleci.com/gh/uktrade/data-hub-frontend/15238) to run `data-hub-frontend` and `datahub-leeloo` through the `SSO` process.

### Of note
This work introduces a new environment variable `OAUTH2_BYPASS_SSO`. This is used in conjunction with `OAUTH2_DEV_TOKEN` for developers who wish to byPass SSO locally.

### CODE REVIEW

File: src/middleware/sso-bypass.js [hunk 1]
```diff
@@ -2,17 +2,18 @@ const { get } = require('lodash')
 const config = require('../../config')
 
 /**
- * If a development bearer token has been provided then assign that to req.session.token
- * so SSO functionality can be bypassed for developers.
+ * If a developer wishes to bypass SSO via config and has also provided a development bearer token, then assign
```
Comment (ztolley) [inline src/middleware/sso-bypass.js]: Would you ever bypass SSO without a token or would you ever provide a token on its own?

File: test/unit/middleware/ssoBypass.test.js [hunk 2]
```diff
@@ -12,23 +14,52 @@ describe('SSO bypass middleware', () => {
     this.nextSpy = sandbox.spy()
   })
 
-  describe('without oauth bypass token', () => {
-    it('should call the next middleware without setting the session token', () => {
+  context('without oauth developer token', () => {
```
Comment (ztolley) [inline test/unit/middleware/ssoBypass.test.js]: 👍

File: src/apps/oauth/controllers.js [hunk 3]
```diff
@@ -5,8 +5,7 @@ const uuid = require('uuid')
 const { get, set } = require('lodash')
 
 const config = require('./../../../config')
-// @TODO remove this once we've diagnosed the cause of the mismatches
-const logger = require('./../../../config/logger')
+const logger = require('./../../../config/logger') // @TODO remove this once we've diagnosed the cause of the mismatches
```
Comment (ztolley) [inline src/apps/oauth/controllers.js]: Does the END @TODO not need a matching BEGIN ? I don't use intellij so don't know if this is an IDE thing. If so then it should go above the line no?

File: README.md [hunk 4]
```diff
@@ -178,16 +180,35 @@ or Visual Studio Code.
...
 ```
 
-#### OAuth
+### OAuth
 Data hub uses [uktrade/staff-sso](https://github.com/uktrade/staff-sso) for OAuth. Details of the required environment
-variables needed to setup OAuth can be seen in the env section of this README. For further information about how to 
-setup OAuth please speak to the [#technology-sso](https://ditdigitalteam.slack.com/messages/C5FLP2DSM/details/) team.
+variables needed to setup OAuth can be seen in the [Environment Variables](#environment-variables) section.
+For further information about how to setup OAuth:
+- Look in confluence for details on setting up SSO for developers. Instructions can be found in
```
Comment (ztolley) [inline README.md]: Is it ok to put this information on a public repo and the slack url? Also this is no use to anyone looking at the code not involved in the project.

Comment (ztolley) [reply README.md]: This is useful for the other devs who haven't done much with SSO much. It may also be a good idea to give a quick 10 min talk on how it all works face to face

File: src/middleware/sso-bypass.js (hunk 1 above)
Comment (feedmypixel) [reply src/middleware/sso-bypass.js]: > Would you ever bypass SSO without a token

No if your Bypassing SSO then you need to provide a token as SSO provides the token.

> would you ever provide a token on its own?

Yes and thats when it is pushed through using `mock-sso` https://github.com/uktrade/data-hub-frontend/pull/1121/files#diff-f4ecd7fceda4a51f2f7d9de1e31c6c4aR70

Comment (feedmypixel) [reply src/middleware/sso-bypass.js]: Have a read through the README updates fully and let me know if it still doesn't make sense

File: README.md (hunk 4 above)
Comment (feedmypixel) [reply README.md]: > Is it ok to put this information on a public repo and the slack url?

Its behind a login. I can move this to confluence if required?

@jim68000 ?

Comment (ztolley) [reply README.md]: It's more a comment about the info means something to us but does it mean anything to someone looking at the project. Doesn't matter really

File: src/middleware/sso-bypass.js (hunk 1 above)
Comment (ztolley) [reply src/middleware/sso-bypass.js]: I skimmed through but I think I need to actually try it to get a feel for if anything is missing from the Readme, but that could happen as a follow up.

I just wanted to be sure that you need 2 different values, and that you couldn't get by just on the presence of one.

File: README.md [hunk 5]
```diff
@@ -178,16 +180,35 @@ or Visual Studio Code.
...
+  `Data Hub team > Technical Documentation > Frontend > SSO for developers`.
+- Speak to the [#technology-sso](https://ditdigitalteam.slack.com/messages/C5FLP2DSM/details/) team.
+
+#### Using SSO when developing
+Developers can also tests SSO functionality locally by removing the `OAUTH2_DEV_TOKEN` environment variable and making
```
Comment (yeahfro) [inline README.md]: `tests` => `test`

File: README.md [hunk 6]
```diff
@@ -178,16 +180,35 @@ or Visual Studio Code.
...
+sure that `OAUTH2_BYPASS_SSO` is set to false. To use OAuth locally you will then need to setup SSO for the SSO provider
```
Comment (yeahfro) [inline README.md]: `setup` => `set up` (noun vs. verb)

File: README.md [hunk 7]
```diff
@@ -178,16 +180,35 @@ or Visual Studio Code.
...
+you are using.
+
+#### Using a local token and SSO mock server
+If you wish to run SSO locally you can use the [uktrade/mock-sso](https://github.com/uktrade/mock-sso) repo.
+To pass this application through the SSO flow you will need to setup the following environment variables.
```
Comment (yeahfro) [inline README.md]: `setup` => `set up` (noun vs. verb)
